pyqtorch/noise/protocol.py

```diff
@@ -12,43 +14,144 @@ class NoiseProtocol:
     PHASE_DAMPING = "PhaseDamping"
     GENERALIZED_AMPLITUDE_DAMPING = "GeneralizedAmplitudeDamping"
 
-    def __init__(self, protocol: str, options: dict = dict()) -> None:
-        self.protocol: str = protocol
-        self.options: dict = options
 
-    def __repr__(self) -> str:
-        if self.target:
-            return (
-                f"{self.protocol}(prob: {self.error_probability}, "
-                f"target: {self.target})"
-            )
-        return f"{self.protocol}(prob: {self.error_probability})"
+class NoiseProtocol:
+    """
+    Defines a protocol of noisy quantum channels.
+
+    Args:
+        protocol: single NoiseType instance or list of NoiseType instances, or
+            list of (NoiseType, options) tuple. When passing list of tuples, for
+            each NoiseType the options should be a dict containing the "error_probability",
+            and optionally a "target". If no "target" is present, the noise instance
+            will be applied to the same target of the gate it is used on.
+        error_probability: probability of error when passing a single NoiseType
+            or a list of NoiseTypes. Note that all noise types require a single float for the
+            error_probability, while the PauliChannel and GeneralizedAmplitudeDamping require
+            a tuple of error probabilities.
+
+    Examples:
+    ```
+        from pyqtorch.noise import NoiseProtocol, NoiseType
+
+        # Single noise instance
+        protocol = NoiseProtocol(NoiseType.BITFLIP, error_probability = 0.5)
+        protocol = NoiseProtocol.bitflip(error_probability = 0.5)
+
+        # Multiples noise instances with same probability
+        protocol = NoiseProtocol(
+                        [NoiseType.BITFLIP, NoiseType.PHASEFLIP],
+                        error_probability = 0.5
+                    )
+
+        # Multiples noise instances with different options
+        prot0 = NoiseProtocol.bitflip(0.5)
+        prot1 = NoiseProtocol.pauli_channel((0.1, 0.2, 0.7))
+        protocol = NoiseProtocol([prot0, prot1])
+
+    ```
+    """
+
+    def __init__(
+        self,
+        protocol: NoiseType | list[NoiseType] | list[NoiseProtocol],
+        error_probability: tuple[float, ...] | float | None = None,
+        target: int | None = None,
+    ) -> None:
+
+        self._error_probability = error_probability
+        self._target = target
+
+        if isinstance(protocol, NoiseType):
+            self.noise_instances = [(protocol, error_probability, target)]
+        elif isinstance(protocol, list) and isinstance(protocol[0], NoiseType):
+            self.noise_instances = [(p, error_probability, target) for p in protocol]  # type: ignore [misc]
+        elif isinstance(protocol, list) and isinstance(protocol[0], NoiseProtocol):
```

Comment:
I would use a predicate:

```python
is_protocol_list = isinstance(protocol, list)
```

and use it in the two clauses as it gets evaluated only once.